### PR TITLE
start firecracker VM using ansible and vagrant

### PR DESCRIPTION
**Pull request by bennyz**
To run: ansible-playbook -i vagrant_inventory test_playbook.yml

TODO: 
- make the firecracker vm accessible

### CODE REVIEW

File: ansible/roles/base/tasks/install.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    name: "{{ user }}"
+    groups: docker
+    append: yes
+
+- name: start docker
+  become: yes
+  systemd:
+    name: docker
+    enabled: yes
+    state: restarted
```
Comment (matanshukry) [inline ansible/roles/base/tasks/install.yml]: why not started?
```suggestion
    state: started
```

File: ansible/roles/base/tasks/compile.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+- name: check if binary exists
+  stat:
+    path: src/firecracker/build/release/firecracker
+  register: binary_exists
+
+- name: build firecracker from source
+  shell: src/firecracker/tools/devtool build --release
+  when: binary_exists.stat.exists == False
+
+- include: download.yml
```
Comment (matanshukry) [inline ansible/roles/base/tasks/compile.yml]: Don't put it here. The downloaded files are used in the execute.yml, yet they are children of a sibling tree (compile.yml).

File: ansible/roles/base/tasks/execute.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+- name: check if socket exists
+  stat:
+    path: /tmp/firecracker.socket
+  register: socket_exists
```
Comment (matanshukry) [inline ansible/roles/base/tasks/execute.yml]: you're not doing anything with it. Either remove it, or do something.

File: ansible/roles/base/tasks/execute.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+- name: add fs
+  shell: >
+         curl --unix-socket /tmp/firecracker.socket -i \
+         -X PUT 'http://localhost/drives/rootfs' \
+         -H 'Accept: application/json'           \
+         -H 'Content-Type: application/json'     \
+         -d '{
+              "drive_id": "rootfs",
+              "path_on_host": "./hello-rootfs.ext4",
```
Comment (matanshukry) [inline ansible/roles/base/tasks/execute.yml]: Is this the same path as the one above?
"dest: ~/hello-rootfs.ext4"
- Are they even ran by the same user?
- Even if so, is theh path the same?

File: tests/Vagrantfile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+Vagrant.configure("2") do |config|
+  config.vm.box = "centos/7"
+  config.vm.network "private_network", ip: "192.168.50.50", auto_config: true
+  config.ssh.forward_agent = true
+
+  config.vm.provider "libvirt" do |domain|
```
Comment (matanshukry) [inline tests/Vagrantfile]: add alternative for windows? or is libvirt available for windows too?

File: ansible/roles/base/tasks/install.yml (hunk 1 above)
Comment (bennyz) [reply ansible/roles/base/tasks/install.yml]: it seems the addition of $USER to the docker group does not always take without restarting, but maybe my env was dirty so I'll check again

File: ansible/roles/base/tasks/execute.yml (hunk 3 above)
Comment (bennyz) [reply ansible/roles/base/tasks/execute.yml]: yup, I forgot to use it

File: ansible/roles/base/tasks/execute.yml (hunk 4 above)
Comment (bennyz) [reply ansible/roles/base/tasks/execute.yml]: > Is this the same path as the one above?
yes
> Are they even ran by the same user?
yes
> Even if so, is theh path the same?
not sure what you mean, it downloads the kernel and the rootfs and the curl commands plug them to the firecracker VM

File: tests/Vagrantfile (hunk 5 above)
Comment (bennyz) [reply tests/Vagrantfile]: We will not support Windows at the moment

File: ansible/roles/base/tasks/install.yml (hunk 1 above)
Comment (matanshukry) [reply ansible/roles/base/tasks/install.yml]: Why was docker started before this task?

File: ansible/roles/base/tasks/execute.yml (hunk 4 above)
Comment (matanshukry) [reply ansible/roles/base/tasks/execute.yml]: You said it's the same path - but it's not the same path string.
One uses dot, the other tilde. If it *happens* to be the same path that's nice, but why rely on it? why not use the same path string for both? (either dot or tilde)

File: tests/Vagrantfile (hunk 5 above)
Comment (matanshukry) [reply tests/Vagrantfile]: :(

Comment (bennyz) [reply tests/Vagrantfile]: you can simply add a VBox provider :)

File: ansible/roles/base/tasks/execute.yml (hunk 3 above)
Comment (bennyz) [reply ansible/roles/base/tasks/execute.yml]: Added

File: ansible/roles/base/tasks/compile.yml (hunk 2 above)
Comment (bennyz) [reply ansible/roles/base/tasks/compile.yml]: Done

File: ansible/roles/base/tasks/execute.yml (hunk 4 above)
Comment (bennyz) [reply ansible/roles/base/tasks/execute.yml]: Done

File: ansible/roles/start_vm/tasks/execute.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
+- name: check if socket exists
+  stat:
+    path: /tmp/firecracker.socket
+  register: socket_exists
+
+- name: remove socket if exists
+  file:
+    state: absent
+    path: /tmp/firecracker.socket
+  when: socket_exists.stat.exists == True
```
Comment (matanshukry) [inline ansible/roles/start_vm/tasks/execute.yml]: not needed. "Note that absent will not cause file to fail if the path does not exist as the state did not change"

File: tests/vagrant_inventory [hunk 7]
```diff
@@ -0,0 +1,7 @@
+[vagrant]
+192.168.50.50
+
+[vagrant:vars]
+user=vagrant
+ansible_user=vagrant
+ansible_ssh_private_key_file=.vagrant/machines/default/libvirt/private_key
```
Comment (matanshukry) [inline tests/vagrant_inventory]: aren't you missing a prefix? ~ perhaps?

Comment (bennyz) [reply tests/vagrant_inventory]: nope, it's in the same dir as the inventory

File: ansible/roles/setup_host/tasks/install.yml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    append: yes
+
+- name: start docker
+  become: yes
+  systemd:
+    name: docker
+    enabled: yes
+    state: started
+
+- include: download.yml
```
Comment (matanshukry) [inline ansible/roles/setup_host/tasks/install.yml]: No! still bad! install.yml doesn't use download.yml, why would it include it? ..

Comment (bennyz) [reply ansible/roles/setup_host/tasks/install.yml]: As discussed, it's part of the host preparation at the moment, we can remove it later if needed

**Pull request merged**